JS/Lesson20/emrah/exercises.js

```diff
@@ -1,70 +1,132 @@
-    }
-
-    describe() {
-        return `This is a ${this._color} ${this._name} with a side length of ${this._sideLength}.`
-    }
-    area() {
-        return this._sideLength * this._sideLength;
-    }
+  constructor(color, sideLength) {
+    super("square", color);
+    this._sideLength = sideLength; // _ is used to show that the value indise an object is private
+  }
+
+  describe() {
+    return `This is a ${this._color} ${this._name} with a side length of ${this._sideLength}.`;
+  }
+  area() {
+    return this._sideLength * this._sideLength;
+  }
 }
 
-const triangle = new Shape('triangle', 'Blue');
-const square = new Square('Blue', 5);
+const triangle = new Shape("triangle", "Blue");
+const square = new Square("Blue", 5);
 
 console.log(triangle.describe());
 console.log(square.describe());
 console.log(square.area());
 
 console.log(square);
 
+// 3. Create a Vehicle class with make, model, and year properties, and a start() method that logs a message to the console indicating that the vehicle has started. Then create a Car class that extends the Vehicle class and adds a numDoors property. Override the start() method in the Car class to log a different message to the console indicating that the car has started.
 
+class Vehicle {
+  constructor(make, model, year) {
+    this._make = make;
+    this._model = model;
+    this._year = year;
+  }
+  start() {
+    console.log("The vehicle has started");
+  }
+}
+class Car extends Vehicle {
+  constructor(make, model, year, numDoors) {
+    super(make, model, year);
+    this._numDoors = numDoors;
+  }
+  start() {
+    console.log("The car has started");
+  }
+}
 
-
-// 3. Create a Vehicle class with make, model, and year properties, and a start() method that logs a message to the console indicating that the vehicle has started. Then create a Car class that extends the Vehicle class and adds a numDoors property. Override the start() method in the Car class to log a different message to the console indicating that the car has started.
+const myCar = new Car("Nissan", "Qashqai", 2022, 4);
+myCar.start();
 
 // 4. Create a BankAccount class with balance and interestRate properties, and deposit() and withdraw() methods. Add a transactionHistory property that is an array of objects representing each transaction made on the account, with properties for type (either "deposit" or "withdrawal"), amount, and date ({type: "withdrawal", amount: amount, date: new Date().toLocaleDateString()}). Implement deposit() and withdraw() methods that add new transactions to the transactionHistory array. Implement a get transactionHistory() getter method that returns a copy of the transaction history array, and a get currentBalance() getter method that calculates and returns the current balance of the account based on the transaction history.
+
+class BankAccount {
+  constructor(balance, interestRate) {
+    this._balance = balance;
+    this._interestRate = interestRate;
+    this.transactionHistory = [];
```

Comment:
It's important to mark all internal values in a class with underscore (_), it makes your code easier to maintain.
```suggestion
    this._transactionHistory = [];
```